### PR TITLE
Fix #11150: avoid recomputation in Typedecl.check_wellfounded

### LINKED ISSUE #11150: Bad typer complexity (?) when checking that types are not directly recursive
We are hitting what seems to be a problem in the typer (complexity-wise) when "closing" recursive types, i.e. when doing the following operation

```
type 'expr expr = [ ... | `Unop of unop *  'expr ]
type expr' = expr' expr  (* "closing" the type *)
```

The slow operation is the final one. Investigations revealed that nearly all the time is spent in the function `Typedecl.check_well_founded`. Closing types as above does not scale at all with multiple recursive types that themselves have multiple type variables. We have reproduced that on automatically generated code (specifically crafted for this issue), but the initial problem we encountered was on manually written code, which was of a completely reasonable size. At the peak of our experiments, compilation of our main file (4 kLoC, 7 kLoC after PPX expansion) was taking 2.5 minutes, nearly all of them spent closing our main types. This made the edit-compile cycle very tedious, especially since Merlin is completely dead in the water on such code.

The issue can be reproduced from the code generated by compiling [gen.ml](https://github.com/ocaml/ocaml/files/8389166/gen.ml.txt)  and running it with arguments `4 4`, `5 5` and `6 6`. More precisely, the following lines give interesting results on my machine.
```
$ ocamlc gen.ml -o gen
$ ./gen 4 4  > repro.ml  && /usr/bin/time ocamlc -c repro.ml 
0.65user 0.00system 0:00.66elapsed 99%CPU (0avgtext+0avgdata 14228maxresident)k
0inputs+16outputs (0major+1642minor)pagefaults 0swaps
$ ./gen 5 5  > repro.ml  && /usr/bin/time ocamlc -c repro.ml
68.86user 0.25system 1:09.25elapsed 99%CPU (0avgtext+0avgdata 14344maxresident)k
0inputs+16outputs (0major+1916minor)pagefaults 0swaps
$ ./gen 6 6  > repro.ml  && /usr/bin/time ocamlc -c repro.ml 
11556.35user 0.42system 3:12:40elapsed 99%CPU (0avgtext+0avgdata 15676maxresident)k
0inputs+24outputs (0major+2074minor)pagefaults 0swaps
```

So 0.7s, 69s and 3h 12 mins, even though at each step we only add one new type. For reference, [repro.ml](https://github.com/ocaml/ocaml/files/8389153/repro.ml.txt) with n=6 is only 64 lines long.

The problem is very impactful for us, and we would like to help fixing it. However, given the very limited understanding of the OCaml typer in our team, I don't really know what to start looking for. Would you have any advice?

------------------

A remark that may be related. In the `repro.ml` file, all the types are mutually recursive. However, consider the variant below, generated by passing the argument `Rectangle` instead of `Square` in `gen.ml`.

```
type ('t1) t1 = [
  | `C1_1 of 't1
  | `C1_2 of 't1
  | `C1_3 of 't1
  | `C1_4 of 't1k
]
and ('t1, 't2) t2 = [
  | `C2_1 of 't1 * 't2
  | `C2_2 of 't1 * 't2
  | `C2_3 of 't1 * 't2
  | `C2_4 of 't1 * 't2
]
and ('t1, 't2, 't3) t3 = [
  | `C3_1 of 't1 * 't2 * 't3
  | `C3_2 of 't1 * 't2 * 't3
  | `C3_3 of 't1 * 't2 * 't3
  | `C3_4 of 't1 * 't2 * 't3
]
and ('t1, 't2, 't3, 't4) t4 = [
  | `C4_1 of 't1 * 't2 * 't3 * 't4
  | `C4_2 of 't1 * 't2 * 't3 * 't4
  | `C4_3 of 't1 * 't2 * 't3 * 't4
  | `C4_4 of 't1 * 't2 * 't3 * 't4
]

type 'a typed = {node: 'a; loc: int}

type ('t1) t1_typed = ('t1) t1 typed
type ('t1, 't2) t2_typed = ('t1, 't2) t2 typed
type ('t1, 't2, 't3) t3_typed = ('t1, 't2, 't3) t3 typed
type ('t1, 't2, 't3, 't4) t4_typed = ('t1, 't2, 't3, 't4) t4 typed

type t1' = (t1') t1_typed
and t2' = (t1', t2') t2_typed
and t3' = (t1', t2', t3') t3_typed
and t4' = (t1', t2', t3', t4') t4_typed
```

With N sufficiently big (here the generation was done with arguments 4 4), compiling this program starts taking a lot of time. However, observe that the closing operation can be rewritten into
```
type t1' = (t1') t1_typed
type t2' = (t1', t2') t2_typed
type t3' = (t1', t2', t3') t3_typed
type t4' = (t1', t2', t3', t4') t4_typed
```
This makes the typing phase instantaneous. This rewriting can of course be done by the user, but it may be a bit complex because of name capture. So it would be really interesting to have the compiler perform this reorganization internally.

-------------------

Finally a bit of context. The reason we are using such "open" types is that we are writing an Ada -> Infer frontend based on a [Nanopasses](https://nanopass.org/) approach. Except that this is in OCaml, so each type and each transformation is fully typed. We are keeping the types open for the following reason. Assume the four types below

```
type 'expr expr1 = [ ... | `C of 'expr | `D of 'expr ]
type expr1 = expr1' expr1
type 'expr expr2 = [ ... | `C of 'expr | `E of 'expr]
type expr2' = expr2' expr2
```
A translation function from `expr1'` to `expr2'` would be decomposed into two parts:
- a function that takes a `expr2' expr1` (all but the first level of constructors have been translated), and return a `expr2' expr2` (i.e. a `expr2'`). In particular, this function has to remove the constructor `D ...`.
- a function that recurses within the structure of `expr1`, and propagates the translation.

We need the open versions of the types to annotate the boilerplate functions, and the closed versions to check that our transformation functions are fully well-typed. (This is actually a bit of an over-simplification, but roughly conveys the idea.)

### PR DESCRIPTION
**Pull request by garrigue**
This is an attempt at improving performance in Typedecl.check_wellfounded, by avoiding recomputation.
This seems to solve #11150.

### CODE REVIEW

**Comment by garrigue** [comment]
@gasche ping

**Comment by gasche** [comment]
I would like to help by reviewing this, but I have the feeling that I am missing many basic explanations before I can make sense of the code. For a start:

- What is the role of `check_well_founded`, what does it mean to be "well-founded"? Relatedly, what is the role of `check_recursion`? What does well-founded mean when `-rectypes` is passed?
- Out of curiosity: what is the role of `check_well_founded_manifest` vs. `check_well_founded_decl`?
- There are two different exceptions, `Recursive_abbrev` and `Cycle_in_def`, what is their purpose / difference?

I think it would be helpful to store answers to these questions in the source, as documentation comments. I could volunteer to write those documentation comments (as extra commits to this PR for example), if you don't feel comfortable doing it yourself. But I would need to understand the answers first, and reverse-engineering them from the code would be very time-consuming.

**Comment by garrigue** [comment]
I'll try to answer your questions, but I'm not the original author of the code...

> * What is the role of `check_well_founded`, what does it mean to be "well-founded"? Relatedly, what is the role of `check_recursion`? What does well-founded mean when `-rectypes` is passed?

This function checks all cycles in a type definition contain at least an object or polymorphic variant node. In the `-rectypes` case, it just checks that a node does may not expand to itself (creating a "black hole", and potential unsoundness).

> * Out of curiosity: what is the role of `check_well_founded_manifest` vs. `check_well_founded_decl`?

`check_well_founded_manifest` is for the manifest field of a type definition.
`check_well_founded_decl` checks all types appearing in the definition.

> * There are two different exceptions, `Recursive_abbrev` and `Cycle_in_def`, what is their purpose / difference?

`Recursive_abbrev` means that the abbreviation we are defining has an illicit occurrence inside itself.
`Cycle_in_def` is for other illicit cycles discovered while checking the definition.

> I think it would be helpful to store answers to these questions in the source, as documentation comments. I could volunteer to write those documentation comments (as extra commits to this PR for example), if you don't feel comfortable doing it yourself. But I would need to understand the answers first, and reverse-engineering them from the code would be very time-consuming.

I can write it, but it would indeed be good if I'm not the only one to understand this code.

**Comment by gasche** [comment]
I propose a first documentation commit in https://github.com/gasche/ocaml/commit/22bb7587f119b0010c99509e05f7eab5d535c9f7. (I considered pushing to your PR, but I don't have commit rights on your own repository. I can send you a PR on top of your branch?)

More documentation work remains to be done:
- I would like to document the exception; this can be done independently of the PR
- ideally one would document the implementation of `check_well_founded`; either we document the pre-PR state first, and rebase the PR on top of that, or we document the post-PR code directly. Documenting the post-PR code directly sounds more time-efficient, but also harder.

**Comment by gasche** [comment]
I'm still trying to understand the fact that we raise one of two different errors in the failure case. Naively, when we find an ill-founded cycle in the type graph, one would expect that the criterion for raising the more precise `Recursive_abbrev` error is "the root of the abbreviation we are checking is part of the cycle". But:

1. what does "the abbreviation we are checking" mean in the case of mutually-recursive abbreviations?
2. how can one build examples where none of the roots of the mutually-recursive abbreviations are part of the cycle? Initially I tried with something like `type 'a t = ('a t * 'b) as 'b`, but this goes into a different failure that is checked in some other place.

#### Behavior changes

When exploring my question (1) by testing, I found two changes of behavior of this patch compared to stock 5.0:

This one looks like an improvement:

```ocaml
(* previous behavior (5.0) *)
# type 'a t = int * 'a u and 'a u = bool -> 'a t;;
  ^^^^^^^^^^^^^^^^^^^^^^
Error: The definition of t contains a cycle:
       int * 'a u

(* new behavior (this PR) *)
# type 'a t = int * 'a u and 'a u = bool -> 'a t;;
  ^^^^^^^^^^^^^^^^^^^^^^
Error: The type abbreviation t is cyclic
```

On the other hand, this one looks like a regression:

```ocaml
(* previous behavior (5.0) *)
# type 'a t = int * 'a u and 'a u = bool -> 'a u;;                                                                                                                                                                                           
                         ^^^^^^^^^^^^^^^^^^^^^^^
Error: The type abbreviation u is cyclic

(* new behavior (this PR) *)
# type 'a t = int * 'a u and 'a u = bool -> 'a u;;                                                                                                                                                                                            
  ^^^^^^^^^^^^^^^^^^^^^^
Error: The type abbreviation t is cyclic
```

File: typing/typedecl.ml
Comment (gasche) [inline typing/typedecl.ml]: Could this be written as the more precise

```ocaml
match Ctype.try_expand_once_opt env ty with
| ty' -> ...
| exception Ctype.Cannot_expand -> ...
```

Generally in the type-checker code the use of `try..with` makes it hard to trace which subterm we expect to raise an exception.

**Comment by gasche** [comment]
> `check_well_founded_manifest` is for the manifest field of a type definition.
> `check_well_founded_decl` checks all types appearing in the definition.

Two further questions:
1. What is the purpose of this separation?
2. Isn't the type manifest part of "all types appearing in the definition", and does that mean that the manifest is in fact checked twice? Why?

I see the following code:

```ocaml
let check_well_founded_manifest env loc path decl =
  if decl.type_manifest = None then () else
  let args = List.map (fun _ -> Ctype.newvar()) decl.type_params in
  check_well_founded env loc path (Path.same path) (Ctype.newconstr path args)

let check_well_founded_decl env loc path decl to_check =
  let open Btype in
  let it =
    {type_iterators with
     it_type_expr = (fun _ -> check_well_founded env loc path to_check)} in
  it.it_type_declaration it (Ctype.generic_instance_declaration decl)

let transl_type_decl ... =
  ...
  List.iter (fun (id, decl) ->
    check_well_founded_manifest new_env (List.assoc id id_loc_list)
      (Path.Pident id) decl)
    decls;
  let to_check =
    function Path.Pident id -> List.mem_assoc id id_loc_list | _ -> false in
  List.iter (fun (id, decl) ->
    check_well_founded_decl new_env (List.assoc id id_loc_list) (Path.Pident id)
      decl to_check)
    decls;
  ...
```

It looks like `check_well_founded_manifest` runs with a much stricter `to_check` parameter, `(Path.same path)`, which means (I think?) that mutually-recursive definitions are not checked in turn.

Naively I would expect that `check_well_founded_manifest` is redundant, that is, if a type has an ill-founded manifest according to this definition, then the following call to `check_well_founded_decl` (with more expansions in the manifest) will also fail with an error.

**Comment by gasche** [comment]
In fact, if I disable the first `check_well_founded_manifest` part, the only changes to the testsuite are some `Recursive_abbrev` errors that become `Cycle_in_def` errors -- so, ill-typed programs that remain ill-typed. If there is a program where the manifest check changes typability, it is not in the testsuite.

**Comment by gasche** [comment]
We had a meeting with @garrigue and @t6s and @HyunggyuJang today, where we looked at this code again.

I pushed two more documentation commits to my version of the branch (Jacques: I still cannot push to COCTI directly):
https://github.com/gasche/ocaml/commits/fix11150

One thing we realized during our meeting is that, contrarily to what Jacques remembered, the current code of check_well_founded has 100% been written by Jacques, in two attempts in 2014 and then 2016 to fix the same issue, both changes making the code substantially more complex than the previous (buggy) definition.

- "Fix PR#6505: check for cycles inside type definitions", August 2014
  https://github.com/ocaml/ocaml/commit/851c34bcb1efade379fd4c6a1e93368244daaada
- " Fix PR#6505 for real, including -principal case; stricter handling of recursive types in constraints", July 2016
  https://github.com/ocaml/ocaml/commit/848c20f27d1e92df359ef02987cc2dc7d4c54cc7

We missed the 2018 and 2020 summer well-foundedness events, but is July 2022 finally the time to document this code?

File: typing/typedecl.ml
Comment (gasche) [inline typing/typedecl.ml]: (I'm reading this code again as I'm trying to piece together some more documentation.)

In the case where `not to_check` and `p` is not expandable, your implementation is calling first `List.iter (check false ty0 TypeSet.empty) tyl` line 689 and then `List.iter (check root ty0 parents) tyl` line 698. Is it really necessary to have those two checks, isn't the first check subsumed by the second check?

Conversely, in the case where we have `to_check` and also `root`, then if expansion fail I believe that `List.iter (check root ty0 parents) tyl` is called twice (line 688 and 698). Can we avoid this?

**Comment by gasche** [comment]
While thinking about well-formedness I found what appears to be a bug in 4.14 and 5.0, but is fixed in this PR:

```ocaml
type 'a t = 'b constraint 'a = < x : 'b >

type u = < x : u > t
(* Error: The type abbreviation u is cyclic *)
```

This example appears to Stack_Overflow on 4.14 and loop infinitely on 5.0.

@garrigue could you add such an example with constraints in the testsuite ?

**Comment by gasche** [comment]
@garrigue: as an experiment I added

```ocaml
if not root then () else begin
```

as the very first line of the definition
`let rec check ~root ty0 parents ty`
to disable the recursive checks outside the root prefix of
the checked type expressions.

As far as I can tell, the whole testsuite passes just fine.

What is an example of an "internal cycle" that needs
[check_well_founded]  to continue checking under guarding
type constructors?

**Comment by gasche** [comment]
Over at [my local branch for this PR](https://github.com/ocaml/ocaml/compare/trunk...gasche:ocaml:fix11150) I tried to write a definition of what well-foundedness means, as a way to understand what should be the specification of the `check_well_founded` function (before going into the implementation). I think it was a useful exercise, it's (slightly) easier for me to read the code now. I'm posting that documentation here in case people have feedback. (Maybe it's something that we could consider having, in part, in the reference manual.)

```
(* Note: Well-foundedness for OCaml types

   We want to guarantee that all cycles within OCaml types are
   "guarded".

   More precisly, we consider a reachability relation
     "[t] is reachable [guarded|unguarded] from [u]"
   defined as follows:

   - [t1, t2...] are reachable rectypes-guarded from
     [t1 -> t2], [t1 * t2 * ...], and all other built-in
     contractive type constructors.

     (By rectypes-guarded we mean: guarded if -rectypes is set,
      unguarded if it is not set.)

   - [t1, t2...] are reachable guarded from object types
       [< m1 : t1; m2 : t2; ... >]
     or polymorphic variants
       [[`A of t1 | `B of t2 | ...]].

   -  If [(t1, t2...) c] is an (expandable) abbreviation,
      then its expansion is reachable unguarded from it.
      Note that we do not demand that [t1, t2...] be reachable.

   - If [(t1, t2...) c] is not an expandable abbreviation,
     typically an abstract type or a datatype (variant or record)
     then [t1, t2...] are reachable unguarded from it.

   - the relation is transitive and guardedness of a composition
     is the disjunction of each guardedness:
     if t1 is reachable from t2 and t2 is reachable from t3;
     if t1 is guarded in t2 or t2 is guarded in t3, then
     t1 is guarded in t3, otherwise it is ungarded in t3.

   A type [t] is not well-founded if and only if [t] is reachable
   unguarded in [t].

   Notice that, in the case of datatypes, the arguments of
   a parametrized datatype are reachable (they must not contain
   recursive occurrences of the type), but the definition of the
   datatype is not.

      (* well-founded *)
      type t = Foo of u
      and u = t

      (* ill-founded *)
      type 'a t = Foo of 'a
      and u = u t
      > Error: The type abbreviation u is cyclic

   Indeed, in the second example [u] is reachable unguarded in [u t]
   -- its own definition.
*)

(* Note: Forms of ill-foundedness

   Several OCaml language constructs could introduce ill-founded
   types, and there are several distinct checks that forbid different
   sources of ill-foundedness.

   1. Type aliases.

      (* well-founded *)
      type t = < x : 'a > as 'a

      (* ill-founded, unless -rectypes is used *)
      type t = (int * 'a) as 'a
      > Error: This alias is bound to type int * 'a
      > but is used as an instance of type 'a
      > The type variable 'a occurs inside int * 'a

      Ill-foundedness coming from type aliases is detected by the "occur check"
      used by our type unification algorithm. See typetexp.ml.

   2. Type abbreviations.

      (* well-founded *)
      type t = < x : t >

      (* ill-founded, unless -rectypes is used *)
      type t = (int * t)
      > Error: The type abbreviation t is cyclic

      Ill-foundedness coming from type abbreviations is detected by
      [check_well_founded] below.

  3. Recursive modules.

     (* well-founded *)
     module rec M : sig type t = < x : M.t > end = M

     (* ill-founded, unless -rectypes is used *)
     module rec M : sig type t = int * M.t end = M
     > Error: The definition of M.t contains a cycle:
     >        int * M.t

     This is also checked by [check_well_founded] below,
     as called from [check_recmod_typedecl].

  4. Functor application

     A special case of (3) is that a type can be abstract
     in a functor definition, and be instantiated with
     an abbreviation in an application of the functor.
     This can introduce ill-foundedness, so functor applications
     must be checked by re-checking the type declarations of their result.

     module type T = sig type t end
     module Fix(F:(T -> T)) = struct
       (* this recursive definition is well-founded
          as F(Fixed).t contains no reachable type expression. *)
       module rec Fixed : T with type t = F(Fixed).t = F(Fixed)
     end

     (* well-founded *)
     Module M = Fix(functor (M:T) -> struct type t = < x : M.t > end)

     (* ill-founded *)
     module M = Fix(functor (M:T) -> struct type t = int * M.t end);;
     > Error: In the signature of this functor application:
     >   The definition of Fixed.t contains a cycle:
     >   F(Fixed).t
*)
```

**Comment by garrigue** [comment]
> What is an example of an "internal cycle" that needs [check_well_founded] to continue checking under guarding type constructors?

It might be that the second call inside `transl_declaration` removed the need to recurse inside each call...
That would be good news for the complexity.

**Comment by garrigue** [comment]
Following your remark about not needing to recurse on non-root paths, I have simplified the code.
`check_well_founded` now only recurses until it finds a guarded node, or a node that was already visited with sufficient information.
`check_well_founded_manifest` ensures that abbreviations can be expanded without creating a type that recurses on its root illegally.
`check_well_founded_decl` checks that type declarations do not contain illegal cycles inside them; it shares the already visited node information, and should be more efficient than before.

File: typing/typedecl.ml
Comment (garrigue) [reply typing/typedecl.ml]: Done

**Comment by garrigue** [comment]
I tried to include all your suggestions, except for the documentation.
The behavior has been much simplified, so maybe we can come up with something.

Concerning
```ocaml
(* new behavior (this PR) *)
# type 'a t = int * 'a u and 'a u = bool -> 'a u;;                                                                                                                                                                                            
  ^^^^^^^^^^^^^^^^^^^^^^
Error: The type abbreviation t is cyclic
```
the situation is a bit difficult. The cycle is discovered while checking that `t` is well-founded, so it is reported for `t`. I understand that this is a bit counter-intuitive, as the cycle is actually caused by `u`, but it is not incorrect, as `t` is not well-founded (we cannot reach a guard).
If we are not extremely careful we could do the opposite error, i.e. report a cycle in `u` whereas `u` itself contains no cycle.
I suppose one could do something by analyzing the trace in `parents`, but this is another work.

**Comment by gasche** [comment]
Naive question: why don't we check the cycles of all `to_check` neighbors at the same time? They are all nodes of a big (possible disconnected) graph, and we could check them all at once. This would be more efficient (currently the checking time can grow quadratic if we expand the neighbors), and it may also be a natural way to better think about how to report cross-neighbor cycles.

The code looks much simpler, thanks a lot for the rework. I'll try to have a look today.

**Comment by garrigue** [comment]
Actually, this is not a single graph, but rather n different graphs depending on which type you start from.
This is because type abbreviations can be used polymorphically in other types, but can only recurse monomorphically.
So one has to check each type definition individually.
Note that in the second pass, checking type declarations, the `visited` map is now shared when checking all nodes inside a type, so this should be now close to optimal.

File: typing/typedecl.ml
Comment (gasche) [inline typing/typedecl.ml]: I don't think that this call `self.it_type_expr self ty` is correct. The intent is to call the "super" iterator that will recursively traverse all type sub-expressions, but I think that in practice it will call exactly the same function recursively, which will stop as it already visited the node.

See <https://github.com/ocaml/ocaml/blob/efd714b73b41df468f53360a067b54370ab0f98e/typing/typemod.ml#L263> for an example of "super" recursion with type_iterators.

Do you have an idea of a testsuite example that would distinguish between the incorrect and the incorrect behavior here? You mentioned in your discussion that you really want to check non-root local cycles in this code. Is there a known example where this must be done?

Comment (garrigue) [reply typing/typedecl.ml]: Well spotted. Indeed the call should not be to `self.it_type_expr` but `self.it_do_type_expr` (in this case using `super` would be wrong, as `Btype.type_iterators.it_type_expr` uses marks).
However, this looks more complex than that, as just doing it that way causes uncaught exceptions in the testsuite.
Will look into that.

Comment (garrigue) [reply typing/typedecl.ml]: Actually, the errors were just caused by conflicts with installed files. I wish the testsuite ignored those...

I have committed a fixed version. The only difference is a change in error message in `typing-misc/constraints.ml`.
The change looks like a correct consequence, but this is not really a counter-example, as the code would have been rejected on regularity grounds anyway.

**Comment by gasche** [comment]
@garrigue I propose the following change to make the path/ty0 logic more effective.

```diff
 let check_well_founded env loc path to_check visited ty0 =
-  let rec check parents ty =
+  let rec check parents ty0 ty =
     if TypeSet.mem ty parents then begin
       (*Format.eprintf "@[%a@]@." Printtyp.raw_type_expr ty;*)
       if match get_desc ty0 with
@@ -674,17 +674,17 @@ let check_well_founded env loc path to_check visited ty0 =
     match get_desc ty with
     | Tconstr(p, tyl, _) ->
         let to_check = to_check p in
-        if to_check then List.iter (check parents) tyl;
+        if to_check then List.iter (check parents ty0) tyl;
         begin match Ctype.try_expand_once_opt env ty with
-        | ty' -> check parents ty'
+        | ty' -> check parents ty ty'
         | exception Ctype.Cannot_expand ->
-            if not to_check then List.iter (check parents) tyl
+            if not to_check then List.iter (check parents ty0) tyl
         end
     | _ ->
-        Btype.iter_type_expr (check parents) ty
+        Btype.iter_type_expr (check parents ty0) ty
   in
   let snap = Btype.snapshot () in
-  try Ctype.wrap_trace_gadt_instances env (check TypeSet.empty) ty0
+  try Ctype.wrap_trace_gadt_instances env (check TypeSet.empty ty0) ty0
   with Ctype.Escape _ ->
     (* Will be detected by check_recursion *)
     Btype.backtrack snap
```

File: typing/typedecl.ml
Comment (garrigue) [reply typing/typedecl.ml]: BTW, the difference in principal mode comes from two things: one is the different handling of `as` in normal and principal mode, with a loss of sharing in principal, and the second is that, compared to the original approach where the deep recursion was done inside `check_well_founded`, and as result checked the whole expanded graph, this version only checks all nodes in the unexpanded graph. So the extra complexity of mixing the checks may be worth it after all.

**Comment by garrigue** [comment]
Is it really what we want?
This would only report abbreviations as cyclic if if they expand to themselves without going through the expansion of another abbreviation.
For instance
```ocaml
type 'a t = 'a
type u = u t
```
would be reported as `the definition of u contains a cycle 'u t'` rather than `the abbreviation u is cyclic`.

**Comment by gasche** [comment]
I don't know what we want; what's a specification of when to raise each warning? But as far as I can tell the current code never rebinds `path` or `ty0`, so the error-choice code

```ocaml
      if match get_desc ty0 with
      | Tconstr (p, _, _) -> Path.same p path
      | _ -> false
      then raise (Error (loc, Recursive_abbrev (Path.name path)))
      else raise (Error (loc, Cycle_in_def (Path.name path, ty0)))
```

has a behavior that is fully determined by the call to `check_well_founded`, independent of the type traversal. This is probably not what we want either, and I can see on the testsuite that my patch strictly improves the error messages.

File: typing/typedecl.ml
Comment (gasche) [reply typing/typedecl.ml]: Do we understand why we want to check for local cycles after expansions? Intuitively, if an abbreviation is not mutually-recursive (`to_check`), then we can already assume that its definition does not contain local cycles, and there should be no need to check anything about its expansion, only the arguments. A "neighbor" (`to_check`) declaration may introduce cycles, but it will also be checked in turn by `check_well_founded`. So unless I am missing something, the issue you must have in mind is an abbreviation that does not contain any cycle, but whose expansion introduces a cycle. Is that right? Do we have an example?

File: Changes
Comment (gasche) [inline Changes]: You could add a reviewer :-)

**Review by gasche — APPROVED** [review]
After working on #11722, I have become convinced that the error messages generated by the new algorithm are "secretly better" than the previous one, even though they sometimes look worse on paper. (Secretly: to see that they are better, one needs to look at the whole expansions trace, which is what #11722 does.)

The change in error message (and the difficulty to understand it) was my last blocker with this PR, which is now Approved.

There are still some grey areas, some things that should be documented better, some thing that I don't understand, or that I am not sure that I understand.

- Currently there is no documentation of what well-foundedness is or means, or how it can happen in OCaml types. I tried to write some documentation in https://github.com/ocaml/ocaml/pull/11207#issuecomment-1188160932, which should ideally be looked over again and submitted as its own PR after the current one is merged.
- The error messages are still suboptimal in some cases; there are several "error paths" to reveal a cycle, and the algorithm picked an arbitrary one that is longer than the one a human would expect. Maybe this can be improved.
- The difference between `check_well_founded_decl` and `check_well_founded_manifest` remains obscure and should be documented better.
- The code is careful to check for "inner cycles" (cycles that occur through expansions to strict subterms of the type), but it is unclear to me that those inner cycles can actually occur, I don't know of an example (I think there is none in the testsuite and Jacques wasn't sure either). If those "inner cycles" cannot occur, which would be my intuition (but you never know...), then we could simplify the code again noticeably.

All these can be left to follow-up PRs, the current state is already a nice improvement. Thanks!

File: Changes
Comment (garrigue) [reply Changes]: Done

**Comment by garrigue** [comment]
I squashed, rebased, and adjusted test output (remove last expansion when it involves the defined type).
Ready for check.

File: typing/typedecl.ml
Comment (gasche) [inline typing/typedecl.ml:668]: You are changing the definition of `rec_abbrev` in a non-trivial way here. Before we would check if `ty0` starts with a `path` constructor, which is a static condition (`ty0` and `path` never change during the computation) purely depending on the callsite. Now you check if the last type before the cycle is found (not the type that is actually the proof of circularity, but the one before) is a `path` constructor. What is the intuition here, beyond the fact that it looks nice in the tests?

It would be easier for me to think about this change if the first commit correctly committed its test state. I have the impression that the first commit contains the old testsuite results from before you rebased, and then this commit contains the update to the testsuite corresponding to both commits. If the first commit correctly updated the testsuite, then in this one commit we would see the test diff that corresponds to your change here, which is currently not visible anywhere -- I have to guess.

Comment (garrigue) [reply typing/typedecl.ml:668]: For the tests, the problem is that one first has to bootstrap before running them, so you don't get the originals.
I could have added an intermediate commit with the results for the first commit, but they were actually very different from what we had before, in a degraded way, which is why I did further changes first.

Now, for the rationale behind `rec_abbrev`.
The point is that, for an abbreviation, due to the way caching works, we only get the same node when we expand it.
So if you find a cycle, you should check the path before expansion. Before this was done by updating `ty0`, and didn't work very well. But thanks to your trace it is now easy to get the node before expansion.

Comment (gasche) [reply typing/typedecl.ml:668]: Why do we need a bootstrap for this PR? Are the types stored in .cmi modified by the fact that we expanded abbreviations differently during this check?

Comment (gasche) [reply typing/typedecl.ml:668]: (I still have the impression that the testsuite update could be synchronized with the code update for each commit, possibly by doing a bootstrap before running the test and then not committing the bootstrap. At least I would see in each commit what the error messages generated by the new code is.)

Comment (gasche) [reply typing/typedecl.ml:668]: Thanks for the comment on the expansion. It may be worth having some comment in the code to explain this.

Comment (gasche) [reply typing/typedecl.ml:668]: Now that I'm convinced by your explanation, I care less about the commit structure for the testsuite update. (As in: I think it would be better to do the right thing, but I suspect that asking for this may delay the merging of this PR which I am now eagerly waiting for, so I am lowering my expectations.) I would still be curious to know why we need a bootstrap, and welcome a code comment on the `rec_abbrev` stuff.

(Note: if the trace contains a `Contains` construct after the expansion, your code will have `rec_abbrev = false`, do we want to look through the trailing `Contains` moves for the last expansion in this case?)

Comment (garrigue) [reply typing/typedecl.ml:668]: I don't think one needs to look through the trailing `Contains`: the point is that you will get a cycle as soon as you expand an inner occurrence of the defined type. So you get the cycle before going through an extra `Contains`.

**Comment by garrigue** [comment]
Is it ok to merge after rebase, or do you still see work to do ?
For the bootstraps, I usually try to do the right thing, but this is not easy.
Basically, you generally want to commit the bootstrap at the very end, because you will need to update it when you rebase.
However, without the bootstrap you cannot run the testsuite. Adding and committing by hand might work, but this quickly gets cumbersome. Or maybe I should learn to use magit...

**Comment by gasche** [comment]
I went ahead and merged. Thanks!

**Pull request merged**
